File: server/src/game/effects/BuffEffects.ts
import type { ICardEffect } from './ICardEffect.js';
import type { EffectContext } from './EffectContext.js';
import type { Minion } from '../Minion.js';

export class BuffAttackEffect implements ICardEffect {
  constructor(public readonly amount: number) {}

  execute(context: EffectContext): void {
    const target = context.target as Minion | null;
    if (!target) {
      // Không có target → buff toàn bộ friendly minions.
      context.player.getBoard().forEach((m) => m.modifyAttack(this.amount));
      return;
    }
    target.modifyAttack(this.amount);
  }
}

export class BuffHealthEffect implements ICardEffect {
  constructor(public readonly amount: number) {}

  execute(context: EffectContext): void {
    const target = context.target as Minion | null;
    if (!target) {
      context.player.getBoard().forEach((m) => m.modifyHealth(this.amount));
      return;
    }
    target.modifyHealth(this.amount);
  }
}

File: server-java/src/main/java/vn/coincard/server/game/effects/BuffEffects.java
package vn.coincard.server.game.effects;

import vn.coincard.server.game.Minion;

/** Attack and health buff strategies. */
public final class BuffEffects {
  private BuffEffects() {}

  public static class BuffAttackEffect implements CardEffect {
    private final int amount;
    public BuffAttackEffect(int amount) { this.amount = amount; }
    @Override
    public void execute(EffectContext context) {
      if (context.target instanceof Minion m) m.modifyAttack(amount);
    }
  }

  public static class BuffHealthEffect implements CardEffect {
    private final int amount;
    public BuffHealthEffect(int amount) { this.amount = amount; }
    @Override
    public void execute(EffectContext context) {
      if (context.target instanceof Minion m) m.modifyHealth(amount);
    }
  }
}
